A partial re-write of the coordinator/agents.

hrichardlee opened the pull request.

kurtschelfthout commented on examples/covid/pyproject.toml:
```diff
@@ -9,7 +9,7 @@ packages = [{ include = "scripts_redirect.py" }]
-meadowgrid_job_worker_ = 'scripts_redirect:meadowgrid_job_worker_'
+meadowgrid_agent_ = 'scripts_redirect:meadowgrid_agent_'
 meadowflow_ = 'scripts_redirect:meadowflow_'
 
```
Not a new change, but hadn't noticed before - why the ending underscore?

hrichardlee replied:
Okay so this is in the examples/covid project which has an "editable-installed local dependency" on the main meadowdata project. The main meadowdata project has regular `meadowgrid_coordinator`, `meadowgrid_agent`, `meadowflow` scripts, so the examples/covid project should have those same scripts available without any additional work. However, there's a bug in poetry that prevents this from happening just for editable-installed local dependencies: https://github.com/python-poetry/poetry/issues/3265 (link also in a comment). Until that's fixed, I added this whole "scripts_redirect" thing. I.e. once that poetry bug is fixed this entire section will go away, and even now, it only matters for people who want to run these examples with a local version of the meadowdata package.

kurtschelfthout commented on src/protobuf_definitions/meadowgrid/meadowgrid.proto:
```diff
@@ -236,48 +255,47 @@ message ProcessState {
         RUN_REQUESTED = 1;
-        // The meadowgrid coordinator has assigned the Job to a worker
-        // TODO add worker_id
-        ASSIGNED = 2;
```
Not sure if relevant here, but in the context of the scheduler (i.e. meadowflow) I am expecting that we'll need a state "assigned to runner" which stores the runner id + request id. This is because we can't have a "transaction" where the scheduler both stores and requests a job to be run - so instead scheduler decides first what the id is and who is going to run it, stores that, and then asks the runner to run it. Then if the scheduler goes down at any point it can "recover" on startup where it was without running the same job twice (assuming a request to a single job runner to run a task is idempotent, which is what the request id is for).

Perhaps for grid it doesn't matter that much.

hrichardlee replied:
Right now, we have three services involved: meadowflow -> meadowgrid coordinator -> meadowgrid agent. Currently, the workflow is:
- meadowflow creates a request id and stores it (in memory), but does NOT decide which meadowgrid agent will run the job)
- meadowflow sends the request to the meadowgrid coordinator (it's safe to send duplicate requests with the same id, subsequent requests to the coordinator with the same id will be ignored: https://github.com/meadowdata/meadowdata/blob/main/src/protobuf_definitions/meadowgrid/meadowgrid.proto#L156)
- the coordinator decides which agent will run the job and stores that (in memory)
- the agent runs the job
So I think the scenario you're talking about works okay (other than the fact that all the storing is only in memory right now)?

This change doesn't "get rid of" the coordinator keeping track of which jobs are assigned to which agents, it just makes it more complicated, because now there's both "pending" (coordinator has decided on an agent for the job, but the agent doesn't know it yet) and "assigned" (the agent has "picked up" the pending job, but it hasn't started running yet).

hrichardlee committed:
A partial re-write of the coordinator/agents.

The coordinator now assigns jobs to agents (in memory) and then agents come "pick up" their jobs that they've been assigned. Previously, the coordinator had no representation of the full set of agents and would just wait for an agent to request a job and choose a job for the agent at that time. This new implementation allows the coordinator to be a little more strategic in deciding which jobs should go to which agent (see _create_workers_for_job, this addresses a previous TODO), but the real impetus is to prepare for future changes that will allow the coordinator to create new agents (e.g. in AWS/Azure) to run jobs. In order to do this, the coordinator needs to know what the universe of available agents is and it needs to be able to "refrain" from sending jobs to existing agents if we are waiting for new agents for that job to be spun up in AWS/Azure. Related minor changes:
- Renaming job_worker to agent to reduce confusion.
- Agents register themselves with the coordinator
- get_next_job is now get_next_jobs so that the coordinator can give an agent more than one job at a time
- Moved management of resources from the agent to the coordinator
- Fleshed out coordinator's representation of agents/grid workers, and moved data classes and logic into resource_allocation.py
- Get rid of the ASSIGNED state, no longer necessary now that we have a richer representation of the state of agents/grid workers.

A related major change is to address some TODOs relating to grid job failures.
- The first is that previously, if a grid_worker failed to start up (e.g. we can't get the docker container image) the coordinator would get a job update with state=RUN_REQUEST_FAILED that would be ignored. We've addressed this by just failing the next unassigned task in that grid job (see update_grid_job_state).
- Second, previously, if a grid_worker failed while running a task, the coordinator would get a job update with state=NON_ZERO_RETURN_CODE, but that would be ignored. We've addressed this by giving the grid workers ids, and having the coordinator keep track of which grid worker is working on which task, which allows the coordinator to fail the appropriate task. This also sets the stage for implementing grid worker timeouts in the future.
- Third, previously, if there a job's required resources prevented it from running on any agent, the job would not fail and no error messages would be thrown. We've addressed this by having the coordinator just fail every task in the grid job/fail the simple job in this scenario.
